### PR TITLE
[IMP] point_of_sale: load default demo data

### PR DESCRIPTION
**Pull request by rhe-odoo**
When loading the pos without demo data, before we had a wizard that allowed the user to load data before the pos starts. Now, we let the user go to the frontend without preloading products and categories in the database. In the front end, when no products and categories are loaded, we let the user press buttons tu load the default products and categories.

The user can also be redirected to the back end to add new products.

task-id: 3054477

---
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

### CODE REVIEW

File: addons/point_of_sale/static/src/js/Screens/ProductScreen/ProductsWidget.js [hunk 1]
```diff
@@ -146,4 +146,17 @@ export class ProductsWidget extends Component {
...
+        this.env.pos.db.add_categories(categories);
+        this.env.pos._loadProductProduct(products);
+        this.env.pos.has_product_available_in_pos = await this.env.services.orm.search("product.product", [
+            ["available_in_pos", "=", true],
+        ]);
+    }
+
+    _createNewProducts() {
+        window.open('/web#action=point_of_sale.action_client_product_menu', '_blank');
+        this.state.show_reload_message = true;
```
Comment (caburj) [inline addons/point_of_sale/static/src/js/Screens/ProductScreen/ProductsWidget.js:160]: rename to `showReloadMessage`.

File: addons/point_of_sale/static/src/js/Screens/ProductScreen/ProductsWidget.js [hunk 2]
```diff
@@ -146,4 +146,17 @@ export class ProductsWidget extends Component {
             }
         }
     }
+    async _loadDemoDataProducts() {
```
Comment (caburj) [inline addons/point_of_sale/static/src/js/Screens/ProductScreen/ProductsWidget.js:149]: no need for underscore.

File: addons/point_of_sale/static/src/js/Screens/ProductScreen/ProductsWidget.js (hunk 1 above)
Comment (caburj) [reply addons/point_of_sale/static/src/js/Screens/ProductScreen/ProductsWidget.js:160]: rename _createNewProducts to createNewProducts.

File: addons/point_of_sale/static/src/xml/Screens/ProductScreen/ProductsWidget.xml [hunk 3]
```diff
@@ -26,6 +26,13 @@
                     <p t-else="">
                         There are no products in this category.
                     </p>
+                    <div t-if="!env.pos.has_product_available_in_pos.length">
```
Comment (caburj) [inline addons/point_of_sale/static/src/xml/Screens/ProductScreen/ProductsWidget.xml:29]: this.state.hasProducts

File: addons/point_of_sale/static/src/js/Screens/ProductScreen/ProductsWidget.js [hunk 4]
```diff
@@ -146,4 +146,17 @@ export class ProductsWidget extends Component {
...
+        const { products, categories } = await this.orm.call("pos.session", "load_product_frontend", [this.env.pos.pos_session.id]);
+        this.env.pos.db.add_categories(categories);
+        this.env.pos._loadProductProduct(products);
+        this.env.pos.has_product_available_in_pos = await this.env.services.orm.search("product.product", [
+            ["available_in_pos", "=", true],
+        ]);
```
Comment (caburj) [inline addons/point_of_sale/static/src/js/Screens/ProductScreen/ProductsWidget.js:155]: this.state.hasProducts = true;

File: addons/point_of_sale/static/src/js/Screens/ProductScreen/ProductsWidget.js [hunk 5]
```diff
@@ -19,7 +19,7 @@ export class ProductsWidget extends Component {
      */
     setup() {
         super.setup();
-        this.state = useState({ previousSearchWord: "", currentOffset: 0 });
+        this.state = useState({ previousSearchWord: "", currentOffset: 0, show_reload_message: false });
```
Comment (caburj) [inline addons/point_of_sale/static/src/js/Screens/ProductScreen/ProductsWidget.js:22]: include:

```
hasProducts: Object.keys(this.env.pos.db.product_by_id).length > 0
```

**Review by caburj — CHANGES_REQUESTED** [review]

File: addons/point_of_sale/wizard/pos_session_check_product_wizard.xml [hunk 6]
```diff
@@ -1,19 +0,0 @@
-<?xml version="1.0" encoding="utf-8"?>
-<odoo>
-    <record id="view_pos_session_check_product_wizard" model="ir.ui.view">
```
Comment (caburj) [inline addons/point_of_sale/wizard/pos_session_check_product_wizard.xml:3]: Not so sure but I think we should also remove this view from the migration.

File: addons/point_of_sale/models/pos_config.py [hunk 7]
```diff
@@ -526,15 +526,6 @@ def open_ui(self):
         domain = [('available_in_pos', '=', True)]
         if self.limit_categories and self.iface_available_categ_ids:
             domain.append(('pos_categ_id', 'in', self.iface_available_categ_ids.ids))
```
Comment (caburj) [inline addons/point_of_sale/models/pos_config.py:528]: `domain` is unused. Maybe we can remove these lines.

File: addons/point_of_sale/static/src/js/Screens/ProductScreen/ProductsWidget.js [hunk 8]
```diff
@@ -21,7 +21,12 @@ export class ProductsWidget extends Component {
      */
     setup() {
         super.setup();
-        this.state = useState({ previousSearchWord: "", currentOffset: 0 });
+        this.state = useState({
+            previousSearchWord: "",
+            currentOffset: 0,
+            showReloadMessage: false,
+            hasProducts: Object.keys(this.env.pos.db.product_by_id).length > 0
```
Comment (caburj) [inline addons/point_of_sale/static/src/js/Screens/ProductScreen/ProductsWidget.js]: Maybe we can make `hasProducts` a computed field but to keep it simple, it's okay this way. (Keeping this note for the future.)

**Review by caburj — CHANGES_REQUESTED** [review]
I pushed a commit to fix the formatting. Pull so that we can have proper formatting of js file in this PR. I also left some comments to resolve.

**Comment by caburj** [comment]
@robodoo r+

**Comment by Xavier-Do** [comment]
@robodoo r-
This conflicts with https://github.com/odoo/odoo/pull/120992 in the same batch

This one will be easier to merge afterward

**Comment by sdegueldre** [comment]
I've force-pushed to adapt uses of `this.env.pos` to `this.pos.globalState` to make it compatible with the above mentionned PR, this should be functionally equivalent. I've confirmed that when cherry picking both commits on top of the latest master, the error that occured in staging no longer occurs after these adaptations.

@robodoo r+

**Pull request closed without merging**